File: 0002-add-two-numbers/0002-add-two-numbers.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode* result = new ListNode();
        ListNode* temp = result;
        int carry = 0;

        while(l1 != NULL || l2 != NULL){
            int sum = 0;
            if(l1 != NULL){
                sum += l1->val;
                l1 = l1->next;
            }
            if(l2 != NULL){
                sum += l2->val;
                l2 = l2->next;
            }
            sum += carry;
            temp->next = new ListNode(sum % 10);
            temp = temp->next;
            carry = sum / 10;
        }
        if(carry != 0){
            temp->next = new ListNode(carry);
        }
        result = result->next;
        return result;
    }
};

File: 0002-add-two-numbers/0002-add-two-numbers.dart
/**
 * Definition for singly-linked list.
 * class ListNode {
 *   int val;
 *   ListNode? next;
 *   ListNode([this.val = 0, this.next]);
 * }
 */
class Solution {
  ListNode? addTwoNumbers(ListNode? l1, ListNode? l2) {
    ListNode? res = new ListNode();
    int carry = 0;
    ListNode? cur = res;
    while(l1 != null || l2 != null){
        int sum = 0;
        if(l1 != null){
            sum += l1.val;
            l1 = l1.next;
        }
        if(l2 != null){
            sum += l2.val;
            l2 = l2.next;
        }
        sum += carry;
        ListNode? newNode = new ListNode(sum % 10);
        carry = sum ~/ 10;
        cur!.next = newNode;
        cur = cur.next;
    }
    if(carry != 0){
        ListNode? newNode = new ListNode(1);
        cur!.next = newNode;
        cur = cur.next;
    }
    return res.next;
  }
}